ksqldb-engine/src/main/java/io/confluent/ksql/planner/plan/JoinNode.java

```diff
@@ -97,6 +116,8 @@ public JoinNode(
     this.left = requireNonNull(left, "left");
     this.right = requireNonNull(right, "right");
     this.withinExpression = requireNonNull(withinExpression, "withinExpression");
+
+    throwOnKeyMismatch(left, right);
```

Comment:
How come this validation happens here rather than in buildStream() where other validation (matching partition counts) occurs? Is it because we actually prefer to validate in the constructor when possible, but the partition count validation requires a topic client which isn't available until buildStream() is called?

Reply:
Where validation happens currently is... a little undefined.  A lot happens in the Analyzer, but personally I'm like to kill off the Analyzer and use the building of the LogicalPlan to validate thing.  I don't think the Analyzer adds anything.

Failing in the constructor is better than `buildStream` as the former means you can't build an invalid logical plan, where as the latter only means you can't build an invalid physical plan.